metagraph/src/graph/representation/hash/dbg_sshash.cpp

```diff
@@ -0,0 +1,234 @@
+        break;
+    case 'C':
+        ssh_idx = nb.backward_C.kmer_id;
+        break;
+    case 'G':
+        ssh_idx = nb.backward_G.kmer_id;
+        break;
+    case 'T':
+        ssh_idx = nb.backward_T.kmer_id;
+        break;
+    default:
+        break;
+    }
+    return ssh_idx + 1;
+}
+
+void DBGSSHash ::adjacent_outgoing_nodes(node_index node,
+                                         const std::function<void(node_index)> &callback) const {
+    assert(node > 0 && node <= num_nodes());
+    call_outgoing_kmers(node, [&](auto child, char) { callback(child); });
+    
+}
+
+void DBGSSHash ::adjacent_incoming_nodes(node_index node,
+                                         const std::function<void(node_index)> &callback) const {
+    assert(node > 0 && node <= num_nodes());
+    call_incoming_kmers(node, [&](auto parent, char) { callback(parent); });
+}
+
+void DBGSSHash ::call_outgoing_kmers(node_index node,
+                                     const OutgoingEdgeCallback &callback) const {
+    assert(node > 0 && node <= num_nodes());
+
+    auto prefix = get_node_sequence(node).substr(1);
+
+    for (char c : alphabet_) {
+        auto next = kmer_to_node(prefix + c);
+        if (next != npos)
+            callback(next, c);
+    }
+}
+
+
+void DBGSSHash ::call_incoming_kmers(node_index node,
+                                     const IncomingEdgeCallback &callback) const {
+    assert(node > 0 && node <= num_nodes());
+
+    std::string suffix = get_node_sequence(node);
+    suffix.pop_back();
+
+    for (char c : alphabet_) {
+        auto prev = kmer_to_node(c + suffix);
+        if (prev != npos)
+            callback(prev, c);
+    }
+}
+
+size_t DBGSSHash::outdegree(node_index node) const {
+    std::string kmer = DBGSSHash::get_node_sequence(node);
+    sshash::neighbourhood nb = dict_->kmer_forward_neighbours(&kmer[0]);
+    size_t out_deg = bool(nb.forward_A.kmer_id + 1) // change to loop?
+                    + bool(nb.forward_C.kmer_id + 1)
+                    + bool(nb.forward_G.kmer_id + 1)
+                    + bool(nb.forward_T.kmer_id + 1);
+    return out_deg;
+}
+
+bool DBGSSHash::has_single_outgoing(node_index node) const {
+    return DBGSSHash::outdegree(node) == 1;
+}
+
+bool DBGSSHash::has_multiple_outgoing(node_index node) const {
+    return DBGSSHash::outdegree(node) > 1;
+}
+
+size_t DBGSSHash::indegree(node_index node) const {
+    std::string kmer = DBGSSHash::get_node_sequence(node);
+    sshash::neighbourhood nb = dict_->kmer_backward_neighbours(&kmer[0]);
+    size_t in_deg = bool(nb.backward_A.kmer_id + 1) // change to loop?
+                    + bool(nb.backward_C.kmer_id + 1)
+                    + bool(nb.backward_G.kmer_id + 1)
+                    + bool(nb.backward_T.kmer_id + 1);
+    return in_deg;
+}
+
+bool DBGSSHash::has_no_incoming(node_index node) const {
+    return DBGSSHash::indegree(node) == 0;
+}
+
+bool DBGSSHash::has_single_incoming(node_index node) const {
```

Comment:
The default implementation involves computing the indegree and checking if it's equal to 1. I feel like this can be done more efficiently for DBGSSHash